### PR TITLE
[ELF] Support placing .lbss/.lrodata/.ldata after .bss

### PR DESCRIPTION
**Pull request by MaskRay**
https://reviews.llvm.org/D150510 places .lrodata before .rodata to
minimize the number of permission transitions in the memory image.
However, this layout is less ideal for -fno-pic code (which is still
important).

Small code model -fno-pic code has R_X86_64_32S relocations with a range
of `[0,2**31)` (if we ignore the negative area). Placing `.lrodata`
earlier exerts relocation pressure on such code. Non-x86 64-bit
architectures generally have a similar `[0,2**31)` limitation if they
don't use PC-relative relocations.

If we place .lrodata later, we will need one extra PT_LOAD. Two layouts
are appealing:

* .bss/.lbss/.lrodata/.ldata (GNU ld)
* .bss/.ldata/.lbss/.lrodata

The GNU ld layout has the nice property that there is only one BSS
(except .tbss/.relro_padding). Add -z lrodata-after-bss to support
this layout.

Since a read-only PT_LOAD segment (for large data sections) may appear
after RW PT_LOAD segments. The placement of `_etext` has to be adjusted.

### CODE REVIEW

File: lld/ELF/Writer.cpp [hunk 1]
```diff
@@ -1124,11 +1127,15 @@ template <class ELFT> void Writer<ELFT>::setReservedSymbolSections() {
   }
 
   if (last) {
-    // _edata points to the end of the last mapped initialized section.
+    // _edata points to the end of the last mapped initialized section before
```
Comment (smithp35) [inline lld/ELF/Writer.cpp:1130]: Apart from the comment about .ldata may be after _edata, is this something we had wrong prior to this patch?

If so it could be worth splitting out into a separate patch so we can identify/test it separately. Ideally we'd want a test case for each of the conditions below.

**Review by smithp35 — COMMENTED** [review]
Matching GNU ld seems sensible to me, although best get this reviewed by someone with an X86_64 background.

File: lld/ELF/Writer.cpp (hunk 1 above)
Comment (aeubanks) [reply lld/ELF/Writer.cpp:1130]: +1

File: lld/ELF/Writer.cpp [hunk 2]
```diff
@@ -988,10 +991,10 @@ static unsigned getSectionRank(OutputSection &osec) {
       osec.relro = true;
     else
       rank |= RF_NOT_RELRO;
-    // Place .ldata and .lbss after .bss. Making .bss closer to .text alleviates
-    // relocation overflow pressure.
+    // Place .lbss/.lrodata/.ldata after .bss. .bss/.lbss being adjacent reuses
+    // the NOBITS size optimization.
     if (osec.flags & SHF_X86_64_LARGE && config->emachine == EM_X86_64)
-      rank |= RF_LARGE;
+      rank |= osec.type == SHT_NOBITS ? 1 : RF_LARGE;
```
[the remark this answers is not in the document]
Comment (MaskRay) [reply lld/ELF/Writer.cpp:997]: I think we can treat `RF_*` as defining larger regions and integer immediates as a mechanism to pick a subregion within a larger region. The immediate use here is similar to `.interp/SHT_NOTE` and target-specific tunings (`.sdata`, `.got`, etc).

File: lld/ELF/Writer.cpp (hunk 1 above)
Comment (MaskRay) [reply lld/ELF/Writer.cpp:1133]: I find that `_etext` needs adjustment and just pushed a commit to fix it.

Hmm. I think splitting this part into a separate PR might be difficult due to the inflexible GitHub PR mechanism.

The test difference is in `lld/test/ELF/x86-64-section-layout.s` below:
```patch
-# CHECK-NEXT: 0000000000205b0a 0 NOTYPE GLOBAL DEFAULT [[#]] (.ldata2) _edata
+# CHECK-NEXT: 0000000000203307 0 NOTYPE GLOBAL DEFAULT [[#]] (.data) _edata
```

File: lld/ELF/Writer.cpp [hunk 3]
```diff
@@ -1103,31 +1106,36 @@ template <class ELFT> void Writer<ELFT>::setReservedSymbolSections() {
...
+    return config->emachine == EM_X86_64 && osec->flags & SHF_X86_64_LARGE;
+  };
   for (Partition &part : partitions) {
     for (PhdrEntry *p : part.phdrs) {
       if (p->p_type != PT_LOAD)
         continue;
       last = p;
-      if (!(p->p_flags & PF_W))
-        lastRO = p;
+      if (!(p->p_flags & PF_W) && p->lastSec && !isLarge(p->lastSec))
```
Comment (MaskRay) [inline lld/ELF/Writer.cpp:1118]: This is a new change. We need to adjust `_etext` as there is now a read-only PT_LOAD segment after RW PT_LOAD segments. (Mentioned in the updated description).

File: lld/ELF/Writer.cpp [hunk 4]
```diff
@@ -1103,31 +1106,36 @@ template <class ELFT> void Writer<ELFT>::setReservedSymbolSections() {
...
+      ElfSym::etext1->section = lastRO;
     if (ElfSym::etext2)
-      ElfSym::etext2->section = lastRO->lastSec;
+      ElfSym::etext2->section = lastRO;
   }
 
   if (last) {
-    // _edata points to the end of the last mapped initialized section.
+    // _edata points to the end of the last mapped initialized section before
+    // the first regular NOBITS section (except .tbss and .relro_padding). In
```
Comment (aeubanks) [inline lld/ELF/Writer.cpp:1134]: if we don't have a .bss section, could this point to .tbss or .relro_padding?

**Review by aeubanks — APPROVED** [review]
lgtm

File: lld/ELF/Writer.cpp (hunk 4 above)
Comment (MaskRay) [reply lld/ELF/Writer.cpp:1134]: Thanks for the catch. After the update the comment no longer makes sense. I should change it to
`// _edata points to the end of the last non-large mapped initialized section.`

I think the primary uses of `_edata` are for linker scripts where our logic here does not matter.
Some code uses something like `p < _edata` to check whether a pointer is guaranteed to be in the read-only segment. It's possible that such code may need updates in case `p` now appears in `.lrodata`. From my experience migrating gold to lld I've only run into one or two instances like this.

**Comment by jyknight** [comment]
Why not condition this on `-no-pic`, and keep the previous more-optimal layout for the normal case?

**Comment by MaskRay** [comment]
> Why not condition this on `-no-pic`, and keep the previous more-optimal layout for the normal case?

The motivation is to avoid layout differences between `-no-pie` and `-pie/-shared`. If someone uses a linker script fragment with `INSERT`, it would be difficult if there are two layouts.

My original objection to doing anything for the `-no-pie` issue (#78521) was because I did not want to pessimize `-pie` users.

I think having two layouts is still an opinion, and I am eager to hear more opinions.

We could also have an option to select between the two layouts and figure out which default it should use. I don't know whether it justifies an option.

**Comment by jyknight** [comment]
ISTM that making the layout conditional in order to not regress the most-common case (PIE) would be worth a minor amount of code in lld, even though the benefit is also relatively small.

Do you think INSERT scripts are going to be a real-world issue there? It seems to me like it wouldn't much matter to someone doing INSERT BEFORE/AFTER .lrodata, if lrodata goes near the beginning or near the end?

**Comment by MaskRay** [comment]
> ISTM that making the layout conditional in order to not regress the most-common case (PIE) would be worth a minor amount of code in lld, even though the benefit is also relatively small.
> 
> Do you think INSERT scripts are going to be a real-world issue there? It seems to me like it wouldn't much matter to someone doing INSERT BEFORE/AFTER .lrodata, if lrodata goes near the beginning or near the end?

This has been my objection to making this layout change at all.
If we consider `-no-pie` an uncommon case that has to be supported, ISTM we can require the user to use a linker script fragment

```
SECTIONS { .lrodata : { *(.lrodata .lrodata.*) } } INSERT AFTER .bss;
```

**Comment by jyknight** [comment]
I don't follow.

no-pie is _uncommon_ but not _unsupported_. Making users add a linker script when they wouldn't otherwise have one, in order to have a working no-pie build seems unjustified.

**Comment by MaskRay** [comment]
The current layout alleviates relocation overflow pressure for `.data` and `.bss` but not `.lrodata`. I don't think this means that we _unsupport_ `-no-pie` use cases. We just don't make `.rodata` nicer to `-no-pie` use cases. We could, but it would be at the cost of layout purity.

Making the layout different depending on `-no-pie`/`-pie` is awkward. `INSERT AFTER .lrodata` would give different behaviors which we should not treat hand-wavy.

**Comment by jyknight** [comment]
> I don't think this means that we unsupport -no-pie use cases

Yes, we'd still support -no-pie, but we'd fail to support -no-pie -mcmodel=medium.

> cost of layout purity

I see that you feel strongly about this (e.g. by calling it "purity"), but I don't understand why. It's not the cost of the "if" in the implementation you're worried about, I think, but some other aspect?

> INSERT AFTER .lrodata would give different behaviors (a 32-bit offset relative to .text is positive in one while negative in the other)

I agree that the sign of the relative offset will change...but why is that concerning? The whole point of "INSERT AFTER" is to avoid specifying the entire layout. Does it matter that the offset relative to .text is sometimes negative and sometimes positive?

> which we should not treat hand-wavy.

Have scripts been broken by the _unconditional_ change from the offset being always-negative to the offset being always-positive? It seems unlikely to me that they would be, but am I missing something? If they don't, then that seems like having it be conditional is also fine.

**Comment by aeubanks** [comment]
+1 to everything jyknight has said. I would prefer `.lrodata` at the beginning of the binary when `-pie` for one less segment, and at the end of the binary when `-no-pie` so large data doesn't increase relocation pressure. This PR is an improvement in that `-no-pie` relocation pressure is significantly more important than one less segment for `-pie` (especially if we already have extra segments due to large data), but not ideal imo.

**Comment by MaskRay** [comment]
> > I don't think this means that we unsupport -no-pie use cases
> 
> Yes, we'd still support -no-pie, but we'd fail to support -no-pie -mcmodel=medium.
>
> > cost of layout purity
> 
> I see that you feel strongly about this (e.g. by calling it "purity"), but I don't understand why. It's not the cost of the "if" in the implementation you're worried about, I think, but some other aspect?
> 
> > INSERT AFTER .lrodata would give different behaviors (a 32-bit offset relative to .text is positive in one while negative in the other)
> 
> I agree that the sign of the relative offset will change...but why is that concerning? The whole point of "INSERT AFTER" is to avoid specifying the entire layout. Does it matter that the offset relative to .text is sometimes negative and sometimes positive?
> 
> > which we should not treat hand-wavy.
> 
> Have scripts been broken by the _unconditional_ change from the offset being always-negative to the offset being always-positive? It seems unlikely to me that they would be, but am I missing something? If they don't, then that seems like having it be conditional is also fine.

"fail to support -no-pie -mcmodel=medium" is an overstatement.
We support `-fpie -mcmodel=medium -no-pie`, and support alleviating relocation overflow pressure for text's references to `.data` and `.bss`.
We just do not alleviate relocation overflow pressure for text's absolute address reference to `.bss` (`-fno-pic`) due to not placing `.lrodata` at the end.

I believe we have some disagreement on (a) how much convenience the linker should provide and (b) whether an uncommon use case should be made explicit (`INSERT [BEFORE|AFTER]`) instead of relying on a linker behavior (especially if the linker code would become ugly).

I think very rarely no-pie users actually care about the codegen difference (e.g. a PIC jump table needs extra instructions).
The motivation is primarily for avoiding dynamic relocations. Have you thought about using `-fpie -no-pie` (or `-fpic`) instead of `-fno-pic -no-pie`?
That would solve the #78521 problem.

When I made the large data layout change, I not only thought about impending needs of x86-64, but also other architectures which have or will get large code model.
ISTM aarch64/ppc64/riscv64 would all be fine with .lrodata at the beginning even in the -fno-pic mode.

**Comment by aeubanks** [comment]
I have heard that x86-64 `-fno-pic` is measurably slower than `-fpie` in large workloads, e.g. having to read RIP when accessing globals RIP-relatively is measurable, as opposed a constant address. There are valid reasons for `-fno-pic`.

It seems unlikely that having to support different layouts for different architectures will impose any noticeable maintenance burden on lld's section sorting code, it's fairly straightforward.

> "fail to support -no-pie -mcmodel=medium" is an overstatement.

I don't think this is an overstatement, lld is not respecting the fact that large sections should not contribute to 32-bit relocation pressure in `-fno-pic` builds, and this actually causes issues in practice.

But again, I'd rather see the current version of this PR go in rather than not go in.

File: lld/ELF/Driver.cpp [hunk 5]
```diff
@@ -1436,6 +1436,8 @@ static void readConfigs(opt::InputArgList &args) {
   config->zInterpose = hasZOption(args, "interpose");
   config->zKeepTextSectionPrefix = getZFlag(
       args, "keep-text-section-prefix", "nokeep-text-section-prefix", false);
+  config->zLrodataAfterBss =
+      getZFlag(args, "lrodata-after-bss", "nolrodata-after-bss", false);
```
Comment (jyknight) [inline lld/ELF/Driver.cpp:1440]: Use `!config->isPic` as default value, instead of false? (moving below the assignment of `isPic`, of course)

Comment (MaskRay) [reply lld/ELF/Driver.cpp:1440]: Sorry, but no. I don't think the x86-64 -fno-pic case justifies changing the tradition that -no-pie/-pie does not change the layout (internal linker script). `-z lrodata-after-bss` is a common ground that we both can accept and it has its own merit to test two layouts for file-backed transparent huge pages.

Comment (jyknight) [reply lld/ELF/Driver.cpp:1440]: OK...then it ought to default to true, instead of false.

(And I do also wish you would explain better why you feel so strongly about not changing behavior based on pic vs not.)

Comment (aeubanks) [reply lld/ELF/Driver.cpp:1440]: default to true makes sense, it works for more cases

Comment (MaskRay) [reply lld/ELF/Driver.cpp:1440]: I have explained. The layout should not magically change if the user switches between -no-pie and -pie/-shared. There are a few options to change the layout (-z keep-text-section-prefix, -z separate-code, --no-rosegment, etc), but none does a "smart" choice. 

With -z separate-code and a large max-page-size, one extra PT_LOAD can have a significant waste. I was actually struggling with the default but now I feel that the current layout makes more sense as the default.

Comment (aeubanks) [reply lld/ELF/Driver.cpp:1440]: > With -z separate-code and a large max-page-size, one extra PT_LOAD can have a significant waste.

This makes sense for the more common `-pie` case.

But *everybody* who's linking with `-no-pie` on x86-64 and has large sections is going to need `.lrodata` at the end of the binary so it doesn't affect small data relocations. This basically forces the user to pass `-z lrodata-after-bss` if they're passing `-no-pie`, which seems like a pointless burden on the user. I don't think people are generally concerned about exactly where `.lrodata` is in the binary, the whole point is that it's *somewhere* far away, doesn't really matter where.

Comment (MaskRay) [reply lld/ELF/Driver.cpp:1440]: I think using "everybody" is not accurate. I've mentioned that `-fpie -no-pie` does not need this. And this `-fno-pie -no-pie` use has been a tiny portion of the user community. We are, after all, dealing with a problem that very few users run into in the first place. I think it matters to optimize for the prevailing configurations (-fpie/-fpic with -pie).

We also need evidence that "I have heard that x86-64 -fno-pic is measurably slower than -fpie in large workloads"

**Review by aeubanks — APPROVED** [review]

File: lld/ELF/Driver.cpp (hunk 5 above)
Comment (MaskRay) [reply lld/ELF/Driver.cpp:1440]: Thanks for understanding. I think I am fairly strong in keeping the stance that layout decisions should not magically change due to -no-pie/-pie. If we ever decide to deviate from the policy, we can re-consider. But having `-z lrodata-after-bss` should make these possible.

**Comment by MaskRay** [comment]
```
Merged
merged 5 commits into [users/MaskRay/spr/main.elf-place-lbsslrodataldata-after-bss](https://github.com/llvm/llvm-project/tree/users/MaskRay/spr/main.elf-place-lbsslrodataldata-after-bss) from [users/MaskRay/spr/elf-place-lbsslrodataldata-after-bss](https://github.com/llvm/llvm-project/tree/users/MaskRay/spr/elf-place-lbsslrodataldata-after-bss)
```

Since the base branch is not main, it seems that "merge" in the web UI merges this to the `/users/MaskRay/spr/...` branch instead of `main`. I guess I need to push this squashed commit to `main` manually.

File: lld/ELF/Driver.cpp (hunk 5 above)
Comment (aeubanks) [reply lld/ELF/Driver.cpp:1440]: When you say "policy" do you mean that the layout policy is documented somewhere?
